### DISCUSSION TITLE
Is it possible to/How can I render a document-string from the commandline without creating the source doc file?

### DISCUSSION THREAD

**Discussion by Gewerd-Strauss** [quarto-dev/quarto-cli#11334] [Q&A]
Hello,

Assume I generate syntax-valid qmd-files programmatically, and then render those to output formats via the quarto commandline. That would simply be

```bash
quarto render doc.qmd --to html [opts]
```

I know that pandoc would allow me to render the file's string directly, that is without having the file itself, e.g.

```bash
printf "## Hello World\nThis is some markdown content." | pandoc -f markdown -t html -o output.html
```

I believe (but am not entirely certain right now) that all format-options which I would usually provide to quarto via a yml-file can be passed on the commandline as well.
This would mean that to my knowledge only the input-document file is strictly required to _be a file_, whereas everything else can be passed directly[^1].

Could this be implemented though? I personally think this to be somewhat odd, but I am probably not aware of some reason or implication against this. 

A potential short-time hack to implement night be to write the string to file, render it via quarto and then remove it again. However, I haven't tested this yet. 

Would this be possible in principal, and would it be something you consider worth implementing, or does this fall outside the scope or design for quarto?

---

I also think that in theory this could allow rendering the same file string to different formats in parallel, but I am not sure to what degree file access and locking would interfere here.

---

Thank you. 
Sincerely, 
~Gw

---

[^1]: Ignoring dependency files like images, data- and code-files, etc... for now.

**Comment by cderv** [comment]
> I believe (but am not entirely certain right now) that all format-options which I would usually provide to quarto via a yml-file can be passed on the commandline as well.
> This would mean that to my knowledge only the input-document file is strictly required to be a file, whereas everything else can be passed directly[1](https://github.com/quarto-dev/quarto-cli/discussions/11334#user-content-fn-1-56d4f7f48f9462a30308ccfb342fb8cd).

To be clear, not all configuration can be passed at command line. Some need to be passed through file, either .yml file, or in document. 

Also, I supposed to you are focusing on single file render, and no project. 

> A potential short-time hack to implement night be to write the string to file, render it via quarto and then remove it again. However, I haven't tested this yet.

It seems a valid workflow to temporarily create what you need. 

> Could this be implemented though? I personally think this to be somewhat odd, but I am probably not aware of some reason or implication against this.

I won't answer on this as I don't know really. I'll let others in the team answer on that.

**Reply by Gewerd-Strauss** [reply]
> Also, I supposed to you are focusing on single file render, and no project.

I think I am only talking about non-project single-document rendering? As in, I haven't spent time thinking about the implications of using this in projects yet; but for single-file I personally would have some use-cases - which is why I am asking to begin with.


> Some need to be passed through file, either .yml file,

Huh well then this defeats the entire point I think. I mean, just going write>render>remove would still be completely valid, but if some configuration options must be specified as a yaml-file, that won't work. 

> or in document.

I assume you are talking about the qmd's frontmatter? But wouldn't that just be part of the file-string? As in, assume quarto could accept 

```bash
printf "## Hello World\nThis is some markdown content." | pandoc -f markdown -t html -o output.html
```

wouldn't


```bash
printf "---\ntitle: "My document"\nformat:\n  html:\n    toc: true\n---\n## Hello World\n This is some markdown content." | quarto whatever the interface would be
```

already be passing the 'in-document'-configuration? (I'm spitballing the commandline-format here, just for elaboration)

**Reply by cderv** [reply]
> But wouldn't that just be part of the file-string?

Yes you're right for anything going in the frontmatter. Project config wouldn't. 

However, I don't really see the use case to be honest. For document with a frontmatter, how would it be simpler to use command line only instead of using a file to store the string ?

Just curious about the full context of such feature you desire.

**Reply by Steinthal** [reply]
@Gewerd-Strauss, I agree with **cderv's** "However, I don't really see the use case to be honest. For document with a frontmatter, how would it be simpler to use command line only instead of using a file to store the string ?"

You could create your ```doc.qmd```with all needed definitions and ```{{< include _mydata.qmd >}}```. Then simply create ```_mydata.qmd```programmatically (overwriting it each time)  and ```quarto render doc.qmd``` (assuming that the desired format is stated in ```doc.qmd```).

If necessary, you could define ```metadata-file: mymeta.yml``` in ```doc.qmd```and create ```mymeta.yml```programmatically as well.

Regarding "I also think that in theory this could allow rendering the same file string to different formats in parallel, but I am not sure to what degree file access and locking would interfere here.":

If you define several formats in ```doc.qmd``` (assuming that you need to define some specifics for each format), you could call ```quarto render doc.qmd --to <formatx>``` for each format defined one after the other.

I use a similar setup for some tasks.

**Reply by Gewerd-Strauss** [reply]
> For document with a frontmatter, how would it be simpler to use command line only instead of using a file to store the string ?

The initial assumption was that all configuration can be provided and declared in the document's frontmatter. Since cderv noted that that was not the case, this whole thing wouldn't be worth it imo. In a perfect world I would like the option to render a document completely from the commandline, without requiring to provide any files at all. 

---

> Just curious about the full context of such feature you desire.

As background:

I write my documents in quarto-syntax using Obsidian.md, and use https://github.com/obsidian-html/obsidian-html to convert obsidian.md's syntax into standard markdown syntax (honestly, I just use it to combine transclusions), e.g.

```md

## A header

![[another note]]
![[another note#section]]

```

This has a couple of advantages over quarto's `{{< include _data.qmd >}}`, namely that I can insert an arbitrary section of a note only; meaning I can write my manuscripts in per-method-notes à la


```md
---
frontmatter
title: method_x
---

# Method X

# Results X 

# Discussion X

# Annex X
```

and insert the related things in my manuscript at the right locations. 

There are more small things which are just faster for me in ObsidianMD than using VScode or Rstudio as an editor. If it's not code, Obsidian is just superior in my opinion, due to very mature plugin environment, templating, hotkeys for just about anything you can think of, and so on....

But to get from here to my rendered manuscript, I must go


1. obsidianMD_note (with mostly quarto-like syntax) > {Obsidian-HTML} > standard_markdown (preserved quarto-like syntax) 
2. \> {own utility for post-processing, final conversion steps I like to do myself} > qmd_file with complete qmd-valid syntax, + generates yaml-config files for each desired output-format, + a variety of other things required 
3. \> {quarto-cli} > rendered outputs

Overall,  I generate 3-4 markdown-flavour files between feeding my original note to ObsidianHTML until I get my rendered outputs from quarto.

I am currently in the process of porting and refactoring my own utility from AHK to python, and am trying to cut down on this pipeline. It is somewhat messy, and became so because this was a growing project over the past 3-or-so-years. And while the ahk-instance is completely fine for my needs _right now_, modifying it becomes tiresome.

ObsidianHTML I can fork and modify myself to return the string of the converted file to stdout if I want, instead of writing it to file.
This means I can eliminate the need for files related to ObsidianHTML's functioning; and the only file-requirements are quarto's document (and yaml-configs) to get to the outputs.

So yea, it's a super specific usecase that makes me ask about this; but I believe that in general terms of automation it could be useful to not require the file itself. I would mod this myself, weren't it that quarto is *_waaaayy_* outside my expertise and I am not even remotely confident in just playing around in it.

---

> Regarding "I also think that in theory this could allow rendering the same file string to different formats in parallel, but I am not sure to what degree file access and locking would interfere here.":
> 
> If you define several formats in `doc.qmd` (assuming that you need to define some specifics for each format), you could call `quarto render doc.qmd --to <formatx>` for each format defined one after the other.


Oh for sure. I do almost the same in my thesis projects, using ahk and R. This works fine, although it takes a bit. Particularly if you have a lot of documents. But doing so in parallel would probably fail due to file-locking if each quarto-process were to target the same source-doc. Just copying the document X times would allow this, but also be against the point of trying to reduce the number of required files.

**Reply by mcanouil** [reply]
Are you aware of the embed shortcode? https://quarto.org/docs/authoring/notebook-embed.html

And contents shortcode? https://prerelease.quarto.org/docs/authoring/contents.html

**Reply by Gewerd-Strauss** [reply]
Yes, but to my understanding
- `{{< contents >}}` only embeds code-cells/divs from a place in documentA into another place in documentA,
- and `{{< embed >}}` only embeds code-cells from `documentA` into `documentB`,

both of which are useful, but don't do what I want/need.

What I do is take an arbitrary chapter of `documentA.md` and insert it into the generated `documentB.md` at an arbitrary place. This allows me to

1. Structure `documentA` as a valid standalone document that I can render on my own, and still
2. take entire sections of it into  `documentB`, anywhere I'd like. With a bit of trickery I can even include the body of a section while omitting the section-header itself.


This makes documentA renderable. I get a single document with the documentation on methods, results and discussion for a specfic method/experiment performed; and then in the overarching submission I can place the relevant sections in the relevant places.

This is a feature of ObsidianHTML.

Additionally, a transcluded section `section` of 

```md
![[note#section]]
```

can itself contain transclusions, with a declared recursion-level which gets resolved. I am not sure if the shortcodes listed by you can resolve text containing these shortcodes themselves ¯\\\_(ツ)_/¯

And finally, doing it via obsidianhtml means I can preview the result via obsidian.md's preview/live-edit modes, as those render these transclusions. It helps sometimes.

Maybe I could set up a similar workflow somehow using quarto alone, but using ObsidianMD is just too convenient for me. I write almost no code in-document for my manuscripts, instead sourcing scripts extensively and only using codeblocks for formatting, figures, tables, and inline-code for, well inline-rendering. This means my writing- and coding-processes are mostly decoupled to different files; instead of doing it all in the document itself. For these reasons, I rarely have a usecase for transcluding codeblocks alone.

**Reply by Steinthal** [reply]
> {{< contents >}} only embeds code-cells/divs from a place in documentA into another place in documentA,

Well, that's not exactly true. You can do the following:

**_documentA-inc.qmd:**
```` qmd

::: {#method-x}
# Method X
Content of method-x

More content…
:::

::: {.content-visible when-meta="documentA"}
# Results X 
:::

::: {#results-x}
Content of results-x
:::


::: {#discussion-x}
# Discussion X
Content of discussion-x

More content…
:::


::: {#annex-x}
# Annex X
Content of annex-x

More content…
:::
````


**documentA.qmd** -- to define the title for the "standalone" document:
```` dmd
---
title: "Doc X"
format: html
documentA: true
---

{{< include _documentA-inc.qmd >}}
````


**documentB.qmd** -- the document combining several sections:
```` qmd
---
title: "Main"
format: html
---

{{< include _documentA-inc.qmd >}}

# Method M
{{< lipsum 1 >}}

{{< contents method-x >}}


# Results M 
{{< lipsum 1 >}}

## Results taken from document X
{{< contents results-x >}}


# Discussion M
{{< lipsum 1 >}}

{{< contents discussion-x >}}


# Annex M
{{< lipsum 1 >}}

{{< contents annex-x >}}
````

## Result

### DocumentA
![Screenshot 2024-11-10 at 22 47 41@2x](https://github.com/user-attachments/assets/d655e027-250c-47e1-b040-3ea147805c3f)


### DocumentB
![Screenshot 2024-11-10 at 22 53 34@2x](https://github.com/user-attachments/assets/a9dab8c1-41e1-4996-8aef-f38f8ef21fba)



Probably still not what you need and maybe too complex, but for me ```{{< contents "dat" >}}``` offers some interesting possibilities.
